### PR TITLE
improve the look of the scripts view

### PR DESCRIPTION
**Pull request by devoncarew**
Improve the look of the scripts view:
- give slightly more room to the scripts view
- don't use horizontal scrolling, but allow horizontal overflow
- make the filter area slightly smaller
- add a badge for the script count in the script area header
- sort the scripts in the debugger controller
- make similar horizontal scrolling changes to the code area
- general work towards https://github.com/flutter/devtools/issues/1648

Screen Shot 2020-04-14 at 4 16 49 PM

Some future improvements are to make the badge slightly smaller (perhaps use https://pub.dev/packages/badges ?), and to remove the dedicated filter area and move it to the script area header.

### CODE REVIEW

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart [hunk 1]
```diff
@@ -210,12 +216,19 @@ class DebuggerScreenBodyState extends State<DebuggerScreenBody>
...
         ),
-        padding: const EdgeInsets.only(left: defaultSpacing),
+        padding: const EdgeInsets.only(left: defaultSpacing, right: 4.0),
         alignment: Alignment.centerLeft,
         height: DebuggerScreen.debuggerPaneHeaderHeight,
-        child: Text(
-          title,
-          style: Theme.of(context).textTheme.subtitle2,
+        child: Row(
+          children: <Widget>[
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart:223]: remove `<Widget>`

File: packages/devtools_app/lib/src/debugger/flutter/scripts.dart [hunk 2]
```diff
@@ -6,85 +6,105 @@ import 'package:flutter/material.dart';
...
+  final void Function(ScriptRef scriptRef) onSelected;
 
   @override
   ScriptPickerState createState() => ScriptPickerState();
 }
 
 class ScriptPickerState extends State<ScriptPicker> {
-  List<ScriptRef> _filtered;
-  TextEditingController filterController = TextEditingController();
+  final TextEditingController _filterController = TextEditingController();
```
Comment (kenzieschmoll) [inline packages/devtools_app/lib/src/debugger/flutter/scripts.dart:33]: remove type on LHS

File: packages/devtools_app/lib/src/debugger/flutter/codeview.dart [hunk 3]
```diff
@@ -114,8 +111,6 @@ class _CodeViewState extends State<CodeView> {
 
   @override
   Widget build(BuildContext context) {
-    // TODO(https://github.com/flutter/devtools/issues/1648): Line numbers,
-    // syntax highlighting and breakpoint markers.
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/codeview.dart:118]: I can re-add this; I think most of the work is also tracked in the github issue.

File: packages/devtools_app/lib/src/debugger/flutter/codeview.dart [hunk 4]
```diff
@@ -162,28 +157,18 @@ class _CodeViewState extends State<CodeView> {
             ),
             const SizedBox(width: denseSpacing),
             Expanded(
-              child: SingleChildScrollView(
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/codeview.dart:165]: I think this is an improvement from a presentation POV in the lists on the left. I'm less certain its a net win for the code area. The issue that I saw here was that we didn't know how wide the text display was, so we had a very large estimate (the longest line in the file * the est. max width of a char). This meant that we overscrolled 2-3x horizontally.

I think I'd prefer not having horizontal scrolling for now (and using ellipsis overflow, w/ the user having the ability to re-size their window and the code area) to the current est. horizontal size behavior.

In the future, we should likely move to a precise horizontal scroll - knowing how long the longest line is in pixels, and allowing users to scroll that far.

File: packages/devtools_app/lib/src/debugger/flutter/common.dart [hunk 5]
```diff
@@ -14,6 +14,26 @@ Widget densePadding(Widget child) {
   );
 }
 
+class Badge extends StatelessWidget {
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/common.dart]: will do

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart [hunk 6]
```diff
@@ -78,18 +78,20 @@ class DebuggerScreenBodyState extends State<DebuggerScreenBody>
       });
     });
     addAutoDisposeListener(controller.currentStack);
-    addAutoDisposeListener(controller.scriptList);
+    addAutoDisposeListener(controller.sortedScripts);
```
[the remark this answers is not in the document]
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart:81]: No, just sortedScripts now.

In future work, we should be more precise about what we invalidate on changes. Currently rebuild more than we need to on stack and breakpoint changes.

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart (hunk 1 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart]: My code completion is very insistent here for some reason.

Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart]: Ah!

@pq, I still see the generic types inserted with the 'wrap with' refactorings (Row, Column, ...).

File: packages/devtools_app/lib/src/debugger/flutter/scripts.dart (hunk 2 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/scripts.dart]: done

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart (hunk 1 above)
Comment (pq) [reply packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart]: Ah!  Missed that.

Looking now (https://github.com/dart-lang/sdk/issues/41517).

File: packages/devtools_app/lib/src/debugger/flutter/codeview.dart (hunk 4 above)
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools_app/lib/src/debugger/flutter/codeview.dart:165]: Can you file a bug to re-enable horizontal scrolling? What you suggested sounds like exactly the right way to do this. If we assume a fixed with font, that should be easy to do efficiently.

File: packages/devtools_app/lib/src/debugger/flutter/scripts.dart [hunk 7]
```diff
@@ -5,86 +5,106 @@
...
   @override
   void didUpdateWidget(ScriptPicker oldWidget) {
     super.didUpdateWidget(oldWidget);
 
-    if (_isNotLoaded) {
-      initFilter();
-    } else if (oldWidget.scripts != widget.scripts) {
-      updateFilter(filterController.text);
-    }
+    updateFilter();
```
Comment (jacob314) [inline packages/devtools_app/lib/src/debugger/flutter/scripts.dart:47]: nit: it would be idiomatic to 
only call updateFilter if
`widget.scripts != oldWidget.scripts`
assuming that object equality is a safe way to check whether the list of scripts changed.

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart:81]: Can you add a TODO for this? Also, can we remove the scriptList Notifier / listener on the controller then?

File: packages/devtools_app/test/flutter/debugger_screen_test.dart [hunk 8]
```diff
@@ -46,6 +46,7 @@ void main() {
       when(debuggerController.breakpoints).thenReturn(ValueNotifier([]));
       when(debuggerController.scriptList)
           .thenReturn(ValueNotifier(ScriptList(scripts: [])));
+      when(debuggerController.sortedScripts).thenReturn(ValueNotifier([]));
```
Comment (jacob314) [inline packages/devtools_app/test/flutter/debugger_screen_test.dart:49]: can we add a test where the sortedScripts list is not empty?

File: packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/debugger_screen.dart:81]: Yup, will do re: the todo.

Re: the `scriptList` Notifier - there aren't any current uses of it. However, it seems useful in terms of listening for basic VM event changes.

I think we'll also need more semantic changes - processed versions of the script lists, processed versions of the breakpoints list. Also things that are closer to UI state - the currently selected script +file location for instance. I'm not sure we want to track all those on the debugger controller, or have it provide more basic info and have another object which tracks post-processed data.

File: packages/devtools_app/lib/src/debugger/flutter/scripts.dart (hunk 7 above)
Comment (devoncarew) [reply packages/devtools_app/lib/src/debugger/flutter/scripts.dart:47]: OK, it looks like this method means that we're re-using the state object from an older statefulwidget.

I don't think object identify will work for us here (wrt the list of scripts). For the future though, we'd need to check both the list of scripts, and the filter controller have the same value? Or, the two fields on the `widget` - the scripts and the `selected` script?

File: packages/devtools_app/test/flutter/debugger_screen_test.dart (hunk 8 above)
Comment (devoncarew) [reply packages/devtools_app/test/flutter/debugger_screen_test.dart:49]: done!
